### DISCUSSION TITLE
Updating Atom in a While Loop for Streaming Text with Throttling

### DISCUSSION THREAD

**Discussion by abdulsamad** [pmndrs/jotai#2292] [Q&A]
I'm working with an API that provides a continuous stream of text data. I'm attempting to update the state within an atom in a while loop. However, I've noticed that the atom only gets updated the first time and the last time (with all the accumulated data). I'm uncertain whether this behavior is due to Jotai optimization, React batching, or another factor. I want to show the streamed text as it gets added (Could be throttled for optimization). Can anyone suggest an alternative approach for updating the state within a while loop, and also recommend a method for incorporating throttling to enhance performance?

I tried using React's `useState` It's working but I'd appreciate it if someone could answer this or what would be the better approach in this case.

The code is something like this 👇

```tsx
const setAtom = useSetAtom(myAtom);

const stream = await getStreamedText();

const reader = stream.getReader();

while (true) {
    const { value, done } = await reader.read();

    if (done) {
          // Stream is completed
         console.log('DONE');
         break;
    }

    setAtom(value);
}

```

**Comment by dai-shi** [comment]
Your code looks good to me. Not sure why it updates only for the first time and the last time.
